Build Ansible Playbooks for each rule in every profile

jan-cerny opened the pull request:
#### Description:
For each profile it creates Ansible Playbooks for each rule that has an Ansible remediation available.  The build is a part of product target for all products.  The Playbook contains values of XCCDF Values selected in the profile. The output is generated to `/build/<product>/playbooks/<profile>`. Each Playbook is in a separate file `<rule_id>.yml`.


#### Rationale:
This format is required by components who need to use Playbooks that contain only fixes for single rule.

pep8speaks commented:
Hello @jan-cerny! Thanks for submitting the PR.



 - In the file [`ssg/yaml.py`](https://github.com/ComplianceAsCode/content/blob/e3e3ad93fbbe7c1953589370504e39ff202c7f35/ssg/yaml.py), following are the PEP8 issues :

> [Line 64:100](https://github.com/ComplianceAsCode/content/blob/e3e3ad93fbbe7c1953589370504e39ff202c7f35/ssg/yaml.py#L64): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (117 > 99 characters)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/8c6b64ac-645d-45e5-866f-807d130f5d59) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/8c6b64ac-645d-45e5-866f-807d130f5d59) completed: **2 new issues, 15 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/c27d27e8-edc3-49e8-8da6-781e19ddbc47) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/c27d27e8-edc3-49e8-8da6-781e19ddbc47) completed: **1 new issues, 15 updated code elements**

yuumasato commented on build-scripts/build_rule_playbooks.py:
```diff
@@ -0,0 +1,272 @@
+
+    def create_playbooks_for_profile(self, profile_path, variables):
+        """
+        Creates playbooks for profile from tasks from snippets.
```
This is confusing.
Maybe you want to say `Creates a playbook for each rule selected in a profile`.

jan-cerny commented on build-scripts/build_rule_playbooks.py:
```diff
@@ -0,0 +1,272 @@
+        self.profiles_dir = os.path.abspath(
+            os.path.join(product_dir, relative_profiles_dir))
+
+    def get_variable_selectors_from_profile(self, profile):
```
Thanks, but I have found this needs to be reworked because `get_rules_from_profile` now ignores extended profiles, which  is wrong.

redhatrises commented:
@jan-cerny can you please be more descriptive for the PR's rationale.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/dcd37bc4-6c4e-420f-8e75-5133a50ace87) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/dcd37bc4-6c4e-420f-8e75-5133a50ace87) completed: **2 new issues, 18 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/99f88cb7-350e-4537-85e2-834026f6b168) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/99f88cb7-350e-4537-85e2-834026f6b168) completed: **2 new issues, 18 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/2fc22b4e-2d86-4268-9981-02d26256217e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/2fc22b4e-2d86-4268-9981-02d26256217e) completed: **1 new issues, 18 updated code elements**

yuumasato commented on build-scripts/build_rule_playbooks.py:
```diff
@@ -0,0 +1,335 @@
+            self.create_playbook(
+                snippet_path, rule_id, variables,
+                profile_refines, profile_playbooks_dir
+            )
```
Please, add an error message if specified `rule_id` is not part of the profile.

jan-cerny replied:
Done.

yuumasato commented on build-scripts/build_rule_playbooks.py:
```diff
@@ -0,0 +1,335 @@
+        profile_playbooks_dir = os.path.join(self.output_dir, profile.id_)
+        os.makedirs(profile_playbooks_dir)
+
+        for snippet in os.listdir(self.input_dir):
```
It should be faster to cycle through `profile_rules` instead of all snippets found in directory.

jan-cerny replied:
Fixed

yuumasato commented on ssg/build_yaml.py:
```diff
@@ -131,6 +131,16 @@ def to_xml_element(self):
 
         return element
 
+    def get_rules(self):
```
`get_rules` can give impression that it returns all rules from profile, while it only returns explicitly selected rules.
The rules selected by "extension" are not returned.
My suggestion: `get_rule_selectors`

jan-cerny replied:
Fixed

yuumasato commented on build-scripts/build_rule_playbooks.py:
```diff
@@ -0,0 +1,335 @@
+                sys.stderr.write(
+                    "Could not parse profile %s.\n" % extended_profile_path)
+                return None
+            rules, variables = self.get_profile_selections(extended_profile)
```
I have a feeling that processing of "extened rules" should also be part of "Profile" class, but not sure if it will be useful.
I'm fine if it remains here though.

jan-cerny replied:
The profile class would need to have context of other profiles. I think it isn't necessary for now.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/c458c5f1-cbcb-4573-92f0-882f3749c4c2) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/c458c5f1-cbcb-4573-92f0-882f3749c4c2) completed: **2 new issues, 21 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/ef94fc74-b8b8-4198-a1d4-94a957e217ab) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/ef94fc74-b8b8-4198-a1d4-94a957e217ab) completed: **2 new issues, 21 updated code elements**

jan-cerny commented on ssg/build_yaml.py:
```diff
@@ -131,6 +131,16 @@ def to_xml_element(self):
+
+    def get_variable_selectors(self):
+        variables = dict()
+        for var_selection in filter(lambda x: "=" in x, self.selections):
```
If there will be 2 different lists in profile YAML files (one for rules and second for values), then we won't need to create weird methods.

yuumasato replied:
I understand, but separating the variable selectors from rule selectors makes experience of crafting profiles a bit worse.
It is very convenient to have the variable selector close to the rule selector in profile yaml.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/366be18b-fd5f-48a9-aa86-3d6c3ea5eeb5) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/366be18b-fd5f-48a9-aa86-3d6c3ea5eeb5) completed: **1 new issues, 19 updated code elements**

yuumasato commented on ssg/playbook_builder.py:
```diff
@@ -158,6 +158,10 @@ def create_playbook(self, snippet_path, rule_id, variables,
+        tags = set()
+        for task in play_tasks:
+            tags |= set(task.pop("tags", []))
+
```
It would be interesting to have a test for this for a rule with multiple tasks. For example Ansible fix for audit rule privileged commands.

yuumasato replied:
Filed https://github.com/ComplianceAsCode/content/issues/4037 regarding this.

matejak commented on build-scripts/build_rule_playbooks.py:
```diff
@@ -0,0 +1,47 @@
+        help="Input directory that contains all Ansible remediations"
+        "snippets for the product we are building. "
+        "e.g. ~/scap-security-guide/build/fedora/fixes/ansible"
+    )
```
If we know the product YAML, we think that we also know where to look for the input directory - it is `build/<product-that-we-know>/fixes/ansible`. This means that we don't have to require this, because we know the default.

jan-cerny replied:
OK, I'm not sure if we will need them when the script is used standalone, so I created an issue as you suggested. See https://github.com/ComplianceAsCode/content/issues/4034

matejak commented on build-scripts/build_rule_playbooks.py:
```diff
@@ -0,0 +1,47 @@
+        help="Output directory to which the rule-based Ansible playbooks "
+        "will be generated. "
+        "e.g. ~/scap-security-guide/build/fedora/playbooks"
+    )
```
Ditto, we know a sensible default for the output directory in a similar way as with the input dir.

jan-cerny replied:
OK, I'm not sure if we will need them when the script is used standalone, so I created an issue as you suggested. See #4034

matejak commented on build-scripts/build_rule_playbooks.py:
```diff
@@ -0,0 +1,47 @@
+    p.add_argument(
+        "--product-yaml", required=True, dest="product_yaml",
+        help="YAML file with information about the product we are building. "
+        "e.g.: ~/scap-security-guide/rhel7/product.yml"
```
We could transfer this to a positional argument, as it is the only argument that is essentially required.

jan-cerny replied:
Depends on resolution of #4034

matejak commented on build-scripts/build_rule_playbooks.py:
```diff
@@ -0,0 +1,47 @@
+    )
+    p.add_argument(
+        "--profile",
+        help="Generate Playbooks only for given Profile ID"
```
I would mention that we accept short profile IDs s.a. `ospp` and I would make it more explicit that this argument can only restrict the selection of profiles, i.e. that all are on by default.

jan-cerny replied:
Done

matejak commented on ssg/playbook_builder.py:
```diff
@@ -0,0 +1,275 @@
+                tasks.append(item)
+        return tasks, values
+
+    def get_tags_from_comments(self, snippet_path):
```
Are we sure that we need this? Enhanced snippets should already contain those.

jan-cerny replied:
Ah, I forgot to remove it, it's not used anywhere.

jan-cerny replied:
I have removed unused method get_tags_from_comments

matejak commented on ssg/playbook_builder.py:
```diff
@@ -0,0 +1,275 @@
+        for task in play_tasks:
+            tags |= set(task.pop("tags", []))
+
+        # play should be a dictionary, but to keep a logical order of elements
```
If keys are added to `OrderedDict` one-by-one, there is no need for the sequence, or for the comment. Although your approach is correct, I would favour explicit assignment.

jan-cerny replied:
But from this code it's obviously visible that that we want to order it.

matejak replied:
It is obvious from the point when `OrderedDict` appears.

jan-cerny replied:
Done

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/1d27836c-e453-499b-8e78-33f6d4965ae1) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/1d27836c-e453-499b-8e78-33f6d4965ae1) completed: **1 new issues, 18 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/4fd86cd2-2e0a-4aad-8b3d-b7795d7d37e7) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/4fd86cd2-2e0a-4aad-8b3d-b7795d7d37e7) completed: **4 new issues, 18 updated code elements**

jan-cerny commented:
@matejak I have removed unused method `get_tags_from_comments`. I have improved help. I have changed the use of the OrderedDict. I have created https://github.com/ComplianceAsCode/content/issues/4034  to track that required options do not need be required.

yuumasato commented:
Great patch @jan-cerny!
